test/pxd_test.cc

```diff
@@ -0,0 +1,396 @@
+		if (write_bytes > 0) {
+			ASSERT_EQ(write_bytes, oh.len);
+			break;
+		}
+
+		ASSERT_EQ(EBUSY, errno);
+	}
+
+	added_ids.erase(dev_id);
+}
+
+TEST_F(PxdTest, device_size)
+{
+	pxd_add_out add;
+	std::string name;
+	int minor, ret;
+	uint64_t dev_size;
+	const uint64_t target_dev_size = 1024 * 1024;
+
+	add.dev_id = 1;
+	add.size = target_dev_size;
+	add.queue_depth = 128;
+	add.discard_size = 4096;
+	dev_add(add, minor, name);
+
+	boost::iostreams::file_descriptor dev_fd(name);
+	ASSERT_GT(dev_fd.handle(), 0);
+
+	ret = ioctl(dev_fd.handle(), BLKGETSIZE64, &dev_size);
+	ASSERT_EQ(0, ret);
+	ASSERT_EQ(dev_size, target_dev_size);
+
+	dev_remove(add.dev_id);
+}
+
+TEST_F(PxdTest, write)
+{
+	struct pxd_add_out add;
+	struct rdwr_in *rdwr = NULL;
+	struct fuse_in_header *in = NULL;
+	struct pxd_rdwr_in *wr = NULL;
+	struct fuse_out_header oh;
+	std::string name;
+	int minor = 0;
+	char msg_buf[write_len * 2];
+	ssize_t read_bytes = 0;
+
+	// Attach a kernel block device (/dev/pxd/pxd1)
+	add.dev_id = 1;
+	add.size = 1024 * 1024;
+	add.queue_depth = 128;
+	add.discard_size = PXD_LBS;
+	dev_add(add, minor, name);
+
+	// Start a thread to perform writes on the attached device
+	std::thread wt(&PxdTest::write_thread, this, name.c_str());
+
+	// Now read in the request from kernel
+	while (1) {
+		int ret = wait_msg(1);
+		ASSERT_EQ(0, ret);
+
+		read_bytes = read(fd, msg_buf, sizeof(msg_buf));
+		rdwr = reinterpret_cast<rdwr_in *>(msg_buf);
+
+		if (rdwr->in.opcode == PXD_WRITE) {
+			read_block(&rdwr->in, reinterpret_cast<pxd_rdwr_in *>(&rdwr->rdwr));
+			break;
+		}
+	}
+
+	// Process the write request
+	wr = reinterpret_cast<pxd_rdwr_in *>(&rdwr->rdwr);
+	ASSERT_EQ(rdwr->in.opcode, PXD_WRITE);
+	ASSERT_EQ(wr->minor, minor);
+	ASSERT_EQ(wr->offset, 0);
+	ASSERT_EQ(wr->size, write_len);
+	
+	// Reply to the kernel
+	oh.len = sizeof(oh);
+	oh.error = 0;
+	oh.unique = rdwr->in.unique;
+	fprintf(stderr, "%s: reply to kernel: status: %d\n", __func__, oh.error);
+	size_t ret = ::write(fd, &oh, sizeof(oh));
+	ASSERT_EQ(sizeof(oh), ret);
+
+	wt.join();
+
+	// Detach block device
+	dev_remove(add.dev_id);
+}
+
+TEST_F(PxdTest, read)
+{
+	struct pxd_add_out add;
+	struct rdwr_in *rdwr = NULL;
+	struct pxd_rdwr_in *rd = NULL;
+	struct fuse_out_header oh;
+	std::string name;
+	int minor = 0;
+	char msg_buf[write_len * 2];
+	ssize_t read_bytes = 0;
+	struct iovec iov[16];
+	int iovcnt = 0;
+	char buf[PXD_LBS];
+
+	// Attach a kernel block device (/dev/pxd/pxd1)
+	add.dev_id = 1;
+	add.size = 1024 * 1024;
+	add.queue_depth = 128;
+	add.discard_size = PXD_LBS;
+	dev_add(add, minor, name);
+
+	// Start a thread to perform reads on the attached device
+	std::thread rt(&PxdTest::read_thread, this, name.c_str());
+
+	// Now read in the request from kernel
+	while (1) {
+		int ret = wait_msg(1);
+		ASSERT_EQ(0, ret);
+
+		read_bytes = read(fd, msg_buf, sizeof(msg_buf));
+		rdwr = reinterpret_cast<rdwr_in *>(msg_buf);
+
+		if (rdwr->in.opcode == PXD_READ) {
```

Comment:
here many reads can come in... because of the reads from host machine on a new block device being made available. After device being added, first run the poller to throw out all IOs and complete it, then proceed with test IO and read it to replay it. That should be the reason for reads having different size than expected.

Reply:
I'll address this in the next change. The issue is not the number of reads but the size. I see only one IO on dev as expected but the size is greater that what I asked for. Maybe some sysfs setting which is causing this, not sure yet.